### ISSUE TITLE
[Bug] EXEC Transformation Service

### ISSUE THREAD

**Issue by shawly** [openhab/openhab#3971]
# Problem
## My Items

I've set up a local udp listener on port 49880, so I can see what gets passed through from my script.
These are my two test items:

```
root@Server:~# cat /etc/openhab/configurations/items/default.items
...
Switch Socket_GA_Living_TV_Test "Test TV" { udp=">[ON:127.0.0.1:49880:EXEC(elro433 10110 11000 on )], >[OFF:127.0.0.1:49880:EXEC(elro433 10110 11000 off )]" }
Switch Socket_GA_Living_AVR_Test "Test AVR" { udp=">[ON:127.0.0.1:49880:'EXEC(elro433 10110 01100 on )'], >[OFF:127.0.0.1:49880:'EXEC(elro433 10110 01100 off )']" }
...
```
## UDP can't use wildcards for commands

One unrelated issue is, I can't use `udp=">[*:127.0.0.1:49880:'EXEC(elro433 10110 01100 %s)']` because it throws an exception:

```
2016-02-04 13:45:11.127 [INFO ] [c.internal.ModelRepositoryImpl] - Refreshing model 'default.items'
2016-02-04 13:45:11.138 [ERROR] [i.internal.GenericItemProvider] - Binding configuration of type 'udp' of item ‘Socket_GA_Living_TV_Test‘ could not be parsed correctly.
org.openhab.model.item.binding.BindingConfigParseException: couldn't create Command from '*'
```

With the EXEC binding this works fine. `exec=">[*:tx433 elro 10110 10000 %2$s 192.168.0.15 49880]"` (Note: tx433 is my current workaround script which uses netcat)
I don't find it very satisfying to have two lines of the exact same code, because I have to edit IP Port and all the arguments twice when I have to change the item.
## The issue with my actual bash script

My bash script (elro433) is located in the /usr/bin folder so I can execute it without the path. I have a check where it counts the arguments and looks if there are exactly three arguments `<masterdip>`, `<slavedip>` and `<on|off>`. If there are less or more arguments, it echos the number of arguments and the actual arguments.

```
root@Server:~# cat /usr/bin/elro433
...
if [ "$#" -lt 3 ]; then
  err "Wrong number of arguments $# ($*).\nUsage: $0 <masterdip> <slavedip> on|off"
fi
...
```

So using it in the shell works perfectly fine and it returns my TXP string which I want to send to my 433MHz LAN Gateway.
But if I use this script via the EXEC Transformation Service my script throws the wrong number of arguments error:

```
2016-02-04 14:00:53.287 [WARN ] [AbstractDatagramChannelBinding] - Received data [2016-02-04T14:00:52+0100]: Wrong number of arguments  7 (10110 01100 on ")], >[OFF:127.0.0.1:49880:EXEC(elro433" 10110 11000 off).
Usage: /usr/bin/elro433 <masterdip> <slavedip> on|off from an undefined remote end /127.0.0.1:54693. We will not process it
```

As you can see, the EXEC Transformation Service ALSO passes through everything AFTER `10110 11000 on` and I don't know why, as far as I can see my syntax is correct.
I have also tried to split the arguments with @@ with no change, also removed the quotation marks.

It also ignores the OFF command completely, it won't send the OFF string when I turn the Switch off...

If I use the exact command which is documented under the TCP Binding wiki entry, I get:

```
 java.io.FileNotFoundException: /etc/openhab/configurations/transform/my.device.map)')], >[OFF:127.0.0.1:49880:'MAP(my.device.map (Datei oder Verzeichnis nicht gefunden)
```

So there is definitely a bug.
##

**Comment by watou** [comment]
> this script via the EXEC Transformation Service my script throws the wrong number of arguments error

Are you using openHAB 1.8.1?  There was a fix that might be related.

> UDP can't use wildcards for commands

I see [in the code](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.tcp/src/main/java/org/openhab/binding/tcp/protocol/internal/ProtocolGenericBindingProvider.java#L191) why this is so, and a pull request would be needed to allow `*` to match any command.

**Comment by shawly** [comment]
> Are you using openHAB 1.8.1? There was a fix that might be related.

Oh lord.. I feel stupid now, this really got fixed with 1.8.1... it took me so long to make a detailed bug report and then this.. holy shit.. thanks...

> I see in the code why this is so, and a pull request would be needed to allow * to match any command.

Yeah I'd probably have to use [this code](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/internal/ExecGenericBindingProvider.java#L251) and compile the udp binding myself, to get my desired wild cards hmm.

**Comment by watou** [comment]
> Yeah I'd probably have to use this code and compile the udp binding myself, to get my desired wild cards hmm.

Perhaps you could make a surgical change and submit a PR to add the feature?  Or supply diffs?

**Comment by shawly** [comment]
I have added the code from the Exec Binding to the TCP Binding provider and created a PR, lets have a look if Jenkins doesn't go wild and then with some luck it gets merged in time.

**Comment by watou** [comment]
Reopening so issue is closed after full resolution.

**Comment by steve-bate** [comment]
@shawly Will you add a link to your PR or reference the issue from the PR comments which will also link them. Thanks!

**Comment by shawly** [comment]
Sure here it is, #3974

**Comment by 9037568** [comment]
Closed via merged PR #3974
